### DISCUSSION TITLE
How to write randomisation with constraints (i.e., pseudorandomisation)?

### DISCUSSION THREAD

**Discussion by IngridYu0211** [jspsych/jsPsych#2755] [Q&A]
Hi, I'd like to ask about pseudorandomisation. My task has 3 independent variables- the levels of partner knowledge (good/poor), levels of own knowledge (good/poor) and context (competitive/non-competitive). I hence have 2*2*2=8 blocks. I varied partner identities so that participant only encountered players B and C in competitive context; and D and E in non-competitive context. 
I want to pseudorandomise the block order so that 
1) participants do not encounter more than 2 competitive or non-competitive blocks consecutively.
2) participants do not encounter more than 2 blocks in a row with the same level of own knowledge (i.e., no more than 2 consecutive blocks of good own knowledge).
3) participants do not encounter the same partner in a row (i.e., not allowing block order like player B + good own knowledge and then player B + poor own knowledge). 

My code currently looks like this without the pseudorandomisation: 
var blockPlayerFactorsA = {
        blockFrame:['competitive'],
        otherIdentity:['playerB', 'playerC'],
        selfSpeed: ['slow','fast']
        }
    var blockPlayerComboA = jsPsych.randomization.factorial(blockPlayerFactorsA, 1);
    
    var blockPlayerFactorsB = {
        blockFrame:['non-competitive'],
        otherIdentity:['playerD', 'playerE'],
        selfSpeed: ['slow','fast']
        }
    var blockPlayerComboB = jsPsych.randomization.factorial(blockPlayerFactorsB, 1);
    
    var allBPCombo= blockPlayerComboA.concat(blockPlayerComboB);
    
    var FPcomboArray= jsPsych.randomization.shuffleNoRepeats(allBPCombo);

for (block_number=2; block_number<=TotalBlockNumber; block_number++) {
  var thisBlockCombo=FPcomboArray[block_number-2];
.... }

Can someone please help me implementing the constraints? How can I do something like 'shuffle again if player identity was the same for two adjacent blocks'? And no more than 2 consecutive competitive/non-competitive blocks? 

Thanks very much in advance!

**Comment by ChristopheBossens** [comment]
To shuffle again each time a constraint is not met, I would implement a while loop that first shuffles the array. Then check the constraints and if one of them is not met the while loop repeats.

```
    let FPcomboArray;
    let validOrder = false;
    
    while(validOrder == false){
      FPcomboArray= jsPsych.randomization.shuffleNoRepeats(allBPCombo);

      validOrder = true;
      for (block_number=1; block_number < FPcomboArray.length; block_number++) {
        // Check player identity
        if (FPcomboArray[block_number-1].otherIdentity == FPcomboArray[block_number].otherIdentity){
          console.log("Two consecutive players detected");
          validOrder = false;
        }

        // Check consecutive competitive blocks
        if(block_number >= 2){
          if( (FPcomboArray[block_number - 2].blockFrame == FPcomboArray[block_number - 1].blockFrame) && 
              (FPcomboArray[block_number - 1].blockFrame == FPcomboArray[block_number].blockFrame)){
                console.log("Consecutive block frames detected");
                validOrder = false;
              }
        }
      }
    }
```

**Reply by IngridYu0211** [reply]
I see, thank you so much!
